packages/SwingSet/src/kernel/kernel.js

```diff
@@ -287,11 +288,12 @@ export default function buildKernel(
       // remove vatID from the list of live vats, and mark for deletion
       kernelKeeper.deleteVatID(vatID);
       kernelKeeper.markVatAsTerminated(vatID);
-      kernelKeeper.removeVatFromSwingStoreExports(vatID);
+      deferred.push(kernelKeeper.removeVatFromSwingStoreExports(vatID));
       for (const kpid of deadPromises) {
         resolveToError(kpid, makeError('vat terminated'), vatID);
       }
     }
+    await Promise.all(deferred);
```

Comment:
This is necessary to preserve the [log ordering in terminate.test.js](https://github.com/Agoric/agoric-sdk/blob/bf1e6453eee701f3e465120a03877cf44d458cc6/packages/SwingSet/test/vat-admin/terminate/terminate.test.js#L498-L504). Alternatively (or additionally), we could update [bootstrap-terminate-with-presence.js](https://github.com/Agoric/agoric-sdk/blob/bf1e6453eee701f3e465120a03877cf44d458cc6/packages/SwingSet/test/vat-admin/terminate/bootstrap-terminate-with-presence.js#L15-L32) to tolerate ordering variation.

Reply:
I am confused about why the simple `await` was not working.

Reply:
Figuring it out was an interesting dive... [vat-vat-admin.js `terminateWithFailure`](https://github.com/Agoric/agoric-sdk/blob/914bee2aa7472039ba230267d72027b940bf8a1c/packages/SwingSet/src/vats/vat-admin/vat-vat-admin.js#L282) ends with a synchronous call to `D(vatAdminDev).terminateWithFailure`, which through further synchronous calls in [device-vat-admin.js](https://github.com/Agoric/agoric-sdk/blob/master/packages/SwingSet/src/devices/vat-admin/device-vat-admin.js#L253) → [kernel/deviceManager.js](https://github.com/Agoric/agoric-sdk/blob/914bee2aa7472039ba230267d72027b940bf8a1c/packages/SwingSet/src/kernel/deviceManager.js#L56) → [kernel.js](https://github.com/Agoric/agoric-sdk/blob/master/packages/SwingSet/src/kernel/kernel.js#L1760) → [kernelSyscall.js](https://github.com/Agoric/agoric-sdk/blob/master/packages/SwingSet/src/kernel/kernelSyscall.js#L201) → [vat-admin-hooks.js](https://github.com/Agoric/agoric-sdk/blob/master/packages/SwingSet/src/kernel/vat-admin-hooks.js#L97)
invokes `void terminateVat(vatID, true, marshalledReason)`, which means that only the synchronous prelude of `terminateVat` has an effect before the result promise of `E(vatAdminNode).terminateWithFailure(reason)` settles and enqueues notifies.

Before this PR, [`kernelKeeper.removeVatFromSwingStoreExports`](https://github.com/Agoric/agoric-sdk/blob/master/packages/SwingSet/src/kernel/state/kernelKeeper.js#L933) was synchronous and therefore could not impede the [following step](https://github.com/Agoric/agoric-sdk/blob/master/packages/SwingSet/src/kernel/kernel.js#L290-L293)'s rejection of promises yet to be decided by the terminating vat (which therefore preceded fulfillment of the `terminateWithFailure` result promise).

But now that this PR is making `removeVatFromSwingStoreExports` async, a naïve `await` thereof would end the synchronous prelude _before_ dead promise rejection, which is what caused the [assertion in terminate.test.js](https://github.com/Agoric/agoric-sdk/blob/914bee2aa7472039ba230267d72027b940bf8a1c/packages/SwingSet/test/vat-admin/terminate/terminate.test.js#L498-L504) to fail. Deferring the `await` restores the expected ordering, although we may want to move it even _further_ down to avoid bumping following logic out of the synchronous prelude, e.g.:
```js
    if (critical) {
      // The following error construction is a bit awkward, but (1) it saves us
      // from doing unmarshaling while in the kernel, while (2) it protects
      // against the info not actually encoding an error, but (3) it still
      // provides some diagnostic information back to the host, and (4) this
      // should happen rarely enough that if you have to do a little bit of
      // extra interpretive work on the receiving end to diagnose the problem,
      // it's going to be a small cost compared to the trouble you're probably
      // already in anyway if this happens.
      panic(`critical vat ${vatID} failed`, Error(info.body));
    } else if (vatAdminRootKref) {
      // static vat termination can happen before vat admin vat exists
      notifyTermination(
        vatID,
        vatAdminRootKref,
        shouldReject,
        info,
        queueToKref,
      );
    } else {
      console.log(
        `warning: vat ${vatID} terminated without a vatAdmin to report to`,
      );
    }

    // worker, if present, needs to be stopped (note that this only applies to ephemeral vats)
    deferred.push(() => vatWarehouse.stopWorker(vatID));

    await Promise.all(deferred);
```

Reply:
I actually would want to go a step further and remove the async nature of this function, but still make it return a promise. However that would release zalgo somewhat, so we should not do that in a hurry now.

Please go ahead with the proposed change above, but use `PromiseAllOrErrors` to avoid swallowing multiple errors.